src/dune_rules/expander.ml

```diff
@@ -306,10 +307,14 @@ let expand_pform_gen ~(context : Context.t) ~bindings ~dir ~source
       | Project_root ->
         Need_full_expander
           (fun t ->
-            Action_builder.return
-              [ Value.Dir (Path.build (Scope.root t.scope)) ])
-      | Cc -> Need_full_expander (fun t -> (cc t).c)
-      | Cxx -> Need_full_expander (fun t -> (cc t).cxx)
+            Memo.Build.return
```

Comment:
Something doesn't feel quite right here. Morally, an `'a Action_builder.t` is a `('a * Dep.Facts.t) Memo.Build.t`, so a `'a Action_builder.t Memo.Build.t` is a nested monad. I had the same issue in #4343 and solved it by "injecting" my `Memo.Build.t` values into `Action_builder.t` via `Action_builder.memo_build`. Could we do the same here?

Reply:
Indeed, I already had to use `Action_builder.memo_build` in a couple of places. I'll see if I can use it here too.

Reply:
Oh, wow, that seems to simplify a lot. Partially undoing my changes :joy: